File: frontend/src/services/ProductService.js
import apiClient from './ApiClient';

class ProductService {
  // 모든 상품 조회
  static async getAllProducts() {
    return await apiClient.getPublic('/products');
  }

  // ID로 상품 조회
  static async getProductById(id) {
    return await apiClient.getPublic(`/products/${id}`);
  }

  // 상품 생성
  static async createProduct(product) {
    return await apiClient.post('/products', product);
  }

  // 상품 수정
  static async updateProduct(id, product) {
    return await apiClient.put(`/products/${id}`, product);
  }

  // 상품 삭제
  static async deleteProduct(id) {
    return await apiClient.delete(`/products/${id}`);
  }

  // 상품명으로 검색
  static async searchProductsByName(name) {
    return await apiClient.getPublic(`/products/search?name=${encodeURIComponent(name)}`);
  }

  // 가격 범위로 검색
  static async getProductsByPriceRange(minPrice, maxPrice) {
    return await apiClient.getPublic(`/products/search?minPrice=${minPrice}&maxPrice=${maxPrice}`);
  }

  static async searchProducts(filters = {}) {
    const params = new URLSearchParams();
    
    if (filters.keyword) params.append('keyword', filters.keyword);
    if (filters.categoryId) params.append('categoryId', filters.categoryId);
    if (filters.minPrice) params.append('minPrice', filters.minPrice);
    if (filters.maxPrice) params.append('maxPrice', filters.maxPrice);
    if (filters.sortBy) params.append('sortBy', filters.sortBy);
    if (filters.sortDirection) params.append('sortDirection', filters.sortDirection);
    
    const queryString = params.toString();
    const url = queryString ? `/products/search?${queryString}` : '/products/search';
    
    return await apiClient.getPublic(url);
  }
}

export default ProductService;

File: backend/src/main/java/com/shoplite/service/ProductService.java
package com.shoplite.service;

import com.shoplite.model.Product;
import com.shoplite.repository.ProductRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class ProductService {

    @Autowired
    private ProductRepository productRepository;

    public List<Product> getAllProducts() {
        return productRepository.findAll();
    }

    public Product getProductById(Long id) {
        Optional<Product> product = productRepository.findById(id);
        if (product.isPresent()) {
            return product.get();
        }
        throw new RuntimeException("상품을 찾을 수 없습니다. ID: " + id);
    }

    public Product createProduct(Product product) {
        return productRepository.save(product);
    }

    public Product updateProduct(Long id, Product productDetails) {
        Product product = getProductById(id);
        product.setName(productDetails.getName());
        product.setDescription(productDetails.getDescription());
        product.setPrice(productDetails.getPrice());
        product.setImageUrl(productDetails.getImageUrl());
        product.setStockQuantity(productDetails.getStockQuantity());
        return productRepository.save(product);
    }

    public void deleteProduct(Long id) {
        Product product = getProductById(id);
        productRepository.delete(product);
    }

    public List<Product> searchProductsByName(String name) {
        return productRepository.findByNameContainingIgnoreCase(name);
    }

    public List<Product> getProductsByPriceRange(Double minPrice, Double maxPrice) {
        return productRepository.findByPriceBetween(minPrice, maxPrice);
    }

    public List<Product> searchProducts(String keyword, Long categoryId, Double minPrice, Double maxPrice, String sortBy, String sortDirection) {
        // 정렬 방향 설정
        Sort.Direction direction = sortDirection.equalsIgnoreCase("asc") ? Sort.Direction.ASC : Sort.Direction.DESC;
        
        // 유효한 정렬 필드인지 확인하고 기본값 설정
        String validSortBy;
        switch (sortBy) {
            case "name":
            case "price":
                validSortBy = sortBy;
                break;
            case "createdAt":
            default:
                validSortBy = "id"; // createdAt 대신 id 사용 (더 안전)
                break;
        }
        
        Sort sort = Sort.by(direction, validSortBy);
        
        // 모든 상품을 가져온 후 필터링 (간단한 구현)
        List<Product> allProducts = productRepository.findAll(sort);
        
        return allProducts.stream()
                .filter(product -> {
                    // 키워드 필터링
                    if (keyword != null && !keyword.trim().isEmpty()) {
                        String lowerKeyword = keyword.toLowerCase();
                        boolean matchesName = product.getName().toLowerCase().contains(lowerKeyword);
                        boolean matchesDescription = product.getDescription().toLowerCase().contains(lowerKeyword);
                        if (!matchesName && !matchesDescription) {
                            return false;
                        }
                    }
                    
                    // 카테고리 필터링
                    if (categoryId != null) {
                        if (product.getCategory() == null || !product.getCategory().getId().equals(categoryId)) {
                            return false;
                        }
                    }
                    
                    // 가격 범위 필터링
                    if (minPrice != null && product.getPrice() < minPrice) {
                        return false;
                    }
                    if (maxPrice != null && product.getPrice() > maxPrice) {
                        return false;
                    }
                    
                    // 활성화된 상품만
                    return product.getIsActive() == null || product.getIsActive();
                })
                .collect(Collectors.toList());
    }
}
